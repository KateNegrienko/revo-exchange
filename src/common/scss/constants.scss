$success: #a8e36c;
$error: #ff7a7a;

$black: #1a1c1f;
$grey: #8b959e;
$white: #fff;
$neutral: #ddd;
$ligth-neutral: #f4f4f4;
$primary: #0075eb;
$primary-hover: lighten($primary, 2%);
$primary-pressed: darken($primary, 2%);
$primary-disabled: $grey;

$elevation-1: 0 4px 8px 0 rgba(16, 23, 40, 0.04);

$border-radius: 5px;
$border-radius-m: 10px;
$border-radius-l: 30px;

$space-s: 5px;
$space-m: 10px;
$space-l: 20px;
$space-xl: 40px;
$space-xxl: 80px;

$breakpoints-screen: (
  "mobile": 440px,
  "tablet": 768px,
  "desktop": 1120px,
  "desktop-large": 1440px,
);

$font-button: (
  "desktop": 20px,
  "tablet": 19px,
  "mobile": 18px,
);
