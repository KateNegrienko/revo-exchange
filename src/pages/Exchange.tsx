import { FC } from "react";

const Exchange: FC = () => <div>Exchange</div>;

export default Exchange;
