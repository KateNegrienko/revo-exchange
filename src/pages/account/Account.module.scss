@import "../../common/scss/constants.scss";
@import "../../common/scss/mixins.scss";

.root {
  padding: $space-m 0;
  .card {
    @include card-layout();
  }
  .header {
    @include header();
  }
}
