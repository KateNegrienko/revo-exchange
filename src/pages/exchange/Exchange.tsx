import { FC, useEffect } from "react";
import { readExchangeRate } from "../../reducers/currency/currency.actions";
import theme from "./Exchange.module.scss";

const Exchange: FC = () => {
  useEffect(() => {
    readExchangeRate();
  }, []);

  return <div className={theme.root}>Exchange</div>;
};

export default Exchange;
