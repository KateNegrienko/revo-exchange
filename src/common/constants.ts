import { IBankAccount } from "../data/Account";

export enum CURRENCIES {
  USD = "USD",
  EUR = "EUR",
  GBP = "GBP",
}

export const DEFAULT_PRICES = {
  destinationPrice: "",
  sourcePrice: "",
};

export const INITIAL_ACCOUNTS: IBankAccount[] = [
  {
    id: CURRENCIES.EUR,
    value: 50.3,
    symbol: "€",
  },
  {
    id: CURRENCIES.GBP,
    value: 100,
    symbol: "£",
  },
  {
    id: CURRENCIES.USD,
    value: 50,
    symbol: "$",
  },
];
