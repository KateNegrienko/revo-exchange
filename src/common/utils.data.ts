import { IRate } from "../data/Rate";
import { CURRENCIES, INITIAL_ACCOUNTS } from "./constants";

export const TEST_RATES: IRate[] = [
  {
    key: CURRENCIES.USD,
    value: CURRENCIES.USD,
    text: CURRENCIES.USD,
    price: 1,
  },
  {
    key: CURRENCIES.EUR,
    value: CURRENCIES.EUR,
    text: CURRENCIES.EUR,
    price: 2,
  },
  {
    key: CURRENCIES.GBP,
    value: CURRENCIES.GBP,
    text: CURRENCIES.GBP,
    price: 3,
  },
];

export const utilsPriceTestData = [
  {
    sourceData: {
      price: 10,
      sourceRate: 2,
      destinationRate: 3,
    },
    expectedDataForSource: {
      destinationPrice: "6.67",
      sourcePrice: "10",
    },
    expectedDataForDestination: {
      destinationPrice: "10",
      sourcePrice: "15",
    },
  },
  {
    sourceData: {
      price: 4,
      sourceRate: 1,
      destinationRate: 3,
    },
    expectedDataForSource: {
      destinationPrice: "1.34",
      sourcePrice: "4",
    },
    expectedDataForDestination: {
      destinationPrice: "4",
      sourcePrice: "12",
    },
  },
];

export const utilsSetPriceTestData = [
  {
    newPriceProps: {
      rates: TEST_RATES,
      price: 10,
      sourceAccount: INITIAL_ACCOUNTS[0],
      destinationAccount: INITIAL_ACCOUNTS[1],
    },
    expectedDataForSource: {
      destinationPrice: "6.67",
      sourcePrice: "10",
    },
    expectedDataForDestination: {
      destinationPrice: "10",
      sourcePrice: "15",
    },
  },
  {
    newPriceProps: {
      rates: TEST_RATES,
      price: 4,
      sourceAccount: INITIAL_ACCOUNTS[2],
      destinationAccount: INITIAL_ACCOUNTS[0],
    },
    expectedDataForSource: {
      destinationPrice: "2",
      sourcePrice: "4",
    },
    expectedDataForDestination: {
      destinationPrice: "4",
      sourcePrice: "8",
    },
  },
];
