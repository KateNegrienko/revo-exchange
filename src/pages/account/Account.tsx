import { FC, useCallback } from "react";
import { shallowEqual, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { IState } from "../../reducers";
import { accountState } from "../../reducers/account/account.types";
import { ROOT } from "../../router/Root.constants";
import theme from "./Account.module.scss";
import AccountCard from "./components/account-card/AccountCard";
import Card from "../../common/card/Card";

const Account: FC = () => {
  const { accounts }: accountState = useSelector(
    (state: IState) => state.account,
    shallowEqual
  );
  const history = useHistory();
  const handleEchangeClick = useCallback(() => {
    history.push(ROOT.EXCHANGE);
  }, [history]);
  return (
    <div className={theme.root}>
      <Card className={theme.card}>
        <h1 className={theme.header}>Bank accounts</h1>
        {accounts.map((card) => (
          <AccountCard card={card} key={card.id} />
        ))}
        <button onClick={handleEchangeClick}>Echange</button>
      </Card>
    </div>
  );
};

export default Account;
