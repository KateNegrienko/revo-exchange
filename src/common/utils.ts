import { IBankAccount } from "../data/Account";
import { IExchangeSource } from "../data/ExchangeSource";
import { INewPriceProps } from "../data/NewPriceProps";
import { IRate } from "../data/Rate";
import { ExchangeAccountType } from "../pages/exchange/components/exchange-account/ExchangeAccount.interface";
import { DataState } from "../reducers/account/account.types";
import { DEFAULT_PRICES } from "./constants";

export const round = (value: number) => {
  return Math.ceil(value * 100) / 100;
};

export const exchangeMapping = (
  { id, value, symbol }: IBankAccount,
  state: DataState
): IBankAccount => {
  switch (id) {
    case state.sourceAccount.id:
      return {
        id,
        symbol,
        value: round(value - Number(state.sourcePrice)),
      };

    case state.destinationAccount.id:
      return {
        id,
        symbol,
        value: round(value + Number(state.destinationPrice)),
      };

    default:
      return { id, value, symbol };
  }
};

export const findRate = (rates: IRate[], rateId: string) => {
  return rates.find(({ key }: IRate) => key === rateId)?.price;
};

export const getPriceWithSourceType = ({
  price,
  sourceRate,
  destinationRate,
}: IExchangeSource) => {
  const destination = round((price / sourceRate) * destinationRate);
  return {
    destinationPrice: destination.toString(),
    sourcePrice: price.toString(),
  };
};

export const getPriceWithDestinationType = ({
  price,
  sourceRate,
  destinationRate,
}: IExchangeSource) => {
  const source = round((price / destinationRate) * sourceRate);

  return {
    destinationPrice: price.toString(),
    sourcePrice: source.toString(),
  };
};

export const setNewPrice = ({
  rates,
  focusInput,
  price,
  sourceAccount,
  destinationAccount,
}: INewPriceProps) => {
  const sourceRate = findRate(rates, sourceAccount.id);
  const destinationRate = findRate(rates, destinationAccount.id);

  if (sourceRate && destinationRate && price > 0) {
    switch (focusInput) {
      case ExchangeAccountType.SOURCE:
        return getPriceWithSourceType({ price, sourceRate, destinationRate });

      case ExchangeAccountType.DESTINATION:
        return getPriceWithDestinationType({
          price,
          sourceRate,
          destinationRate,
        });
    }
  }
  return DEFAULT_PRICES;
};
