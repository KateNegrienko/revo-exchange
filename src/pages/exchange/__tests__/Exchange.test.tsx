import React from "react";
import * as redux from "react-redux";
import { shallow, ShallowWrapper } from "enzyme";
import Exchange from "../Exchange";
import { INITIAL_ACCOUNT_MODEL } from "../../../reducers/account/account.model";

import { TEST_RATES } from "../../../common/utils.data";
import { INITIAL_ACCOUNTS } from "../../../common/constants";

describe("[Exchange component]", () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ ...INITIAL_ACCOUNT_MODEL, rates: TEST_RATES });
    wrapper = shallow(<Exchange />);
  });

  it("Should render properly with only required props ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Should add disable props if sourceAccount == destinationAccount ", () => {
    expect(wrapper.find("Button").prop("disabled")).toBeFalsy();
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({
      ...INITIAL_ACCOUNT_MODEL,
      rates: TEST_RATES,
      sourceAccount: INITIAL_ACCOUNTS[1],
      destinationAccount: INITIAL_ACCOUNTS[1],
    });
    wrapper = shallow(<Exchange />);

    expect(wrapper.find("Button").prop("disabled")).toBeTruthy();
  });

  it("Should add disable props if sourcePrice < sourceAccountValue ", () => {
    expect(wrapper.find("Button").prop("disabled")).toBeFalsy();
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({
      ...INITIAL_ACCOUNT_MODEL,
      rates: TEST_RATES,
      sourcePrice: 7000,
    });
    wrapper = shallow(<Exchange />);

    expect(wrapper.find("Button").prop("disabled")).toBeTruthy();
  });
});
