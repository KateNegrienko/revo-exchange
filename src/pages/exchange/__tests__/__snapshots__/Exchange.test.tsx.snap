// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Exchange component] Should render properly with only required props  1`] = `
<div
  className="root"
>
  <Card
    className="card"
  >
    <h1
      className="header"
    >
      Exchange
    </h1>
    <ExchangeAccount
      account={
        Object {
          "id": "EUR",
          "symbol": "€",
          "value": 50.3,
        }
      }
      onChangeAccount={[Function]}
      onChangePrice={[Function]}
      price=""
      rates={
        Array [
          Object {
            "key": "USD",
            "price": 1,
            "text": "USD",
            "value": "USD",
          },
          Object {
            "key": "EUR",
            "price": 0.82,
            "text": "EUR",
            "value": "EUR",
          },
          Object {
            "key": "GBP",
            "price": 0.73,
            "text": "GBP",
            "value": "GBP",
          },
        ]
      }
      type="source"
    />
    <ExchangeAccount
      account={
        Object {
          "id": "GBP",
          "symbol": "£",
          "value": 100,
        }
      }
      onChangeAccount={[Function]}
      onChangePrice={[Function]}
      price=""
      rates={
        Array [
          Object {
            "key": "USD",
            "price": 1,
            "text": "USD",
            "value": "USD",
          },
          Object {
            "key": "EUR",
            "price": 0.82,
            "text": "EUR",
            "value": "EUR",
          },
          Object {
            "key": "GBP",
            "price": 0.73,
            "text": "GBP",
            "value": "GBP",
          },
        ]
      }
      type="destination"
    />
    <Button
      disabled={false}
      onClick={[Function]}
    >
      Exchange
    </Button>
  </Card>
</div>
`;
